### ISSUE TITLE
poi game area screenshot button not working

### ISSUE THREAD

**Issue by Whitechapel-1888** [poooi/poi#2420]
**The screenshot button on the bottom toolbar of the game area does not work in 10.8**

_poi version:_ 10.8

_OS:_ Windows 11 home (build 22000.469)

In version 10.6, the screenshot button works without issue.
In version 10.7, it sometimes doesn't work on a fresh start with canvas disabled and does not work at all with canvas enabled. A restart sometimes fixes the issue, but only if the canvas option is disabled. 
In version 10.8, the screenshot button does not work at all, regardless of the canvas setting in the options. 

_How to reproduce:_ Using version 10.8, start the game and click on the screenshot button.

**Comment by KochiyaOcean** [comment]
Cannot reproduce. Will be useful if you can provide screenshot about error message (if has any)

**Comment by Whitechapel-1888** [comment]
> Cannot reproduce. Will be useful if you can provide screenshot about error message (if has any)

There is none, clicking the button does nothing. Will the console give some useful output? I am willing to test and upload the result if it does.

**Comment by KochiyaOcean** [comment]
Yes, please provide screenshot of your console if there is something

**Comment by Whitechapel-1888** [comment]
There is indeed an output when pressing the screenshot button, see the red error message:
![image](https://user-images.githubusercontent.com/44060684/153055596-8acbb19b-c9e4-4ac1-8458-dea7666ed470.png)

I tried a few times, just to be sure. Before I downgrade to 10.6 again, is there something the console can output that helps you in narrowing the issue down?

**Comment by KochiyaOcean** [comment]
It may be related to https://github.com/electron/electron/issues/31992

**Comment by KochiyaOcean** [comment]
Can you try following code in console?
```js
((toClipboard) => {
    const { width, height } = getStore('layout.webview')
    const rect = {
      x: 0,
      y: 0,
      width: Math.floor(width * devicePixelRatio),
      height: Math.floor(height * devicePixelRatio),
    }
    getStore('layout.webview.ref')
      .capturePage(rect)
      .then((image) => {
        console.log(image
            .resize({ width: Math.floor(width), height: Math.floor(height) })
            .toDataURL())
      })
  })()
```

**Comment by Whitechapel-1888** [comment]
I hope I did this correctly:
![image](https://user-images.githubusercontent.com/44060684/153400473-4ddf23b9-923c-457a-ba43-abeb69bb9f58.png)

I am not sure if this matters, but I upgrade poi viewer using winget. I am adding this because winget seems to install different versions for some apps (thinking of WSL).

**Comment by KochiyaOcean** [comment]
Can you try the package version in poi.moe?

**Comment by h5nc** [comment]
Not sure if this helps.

After Electron 13, calling `getStore('layout.webview.ref').getWebContents().capturePage()` through console will result in an unresolved promise. This only reproduce on my Mac.
`getStore('layout.webview.ref').capturePage()` works fine for both platforms.

Then I find out that all promises sending by any Electron webContens, through console or renderer processes, will remain unresolved.
I tried building electron from source with test functions that resolve promises immediately. Calling those functions still returns unresolved handlers.

However in main process `webContents.capturePage()` works fine.

So in my case I added this to app.js
```
 ipcMain.handle('displays::capture-page', async (e) => {
    const image = await mainWindow.webContents.capturePage()
    return image
  })
```

then I can call anywhere with

```
const { ipcRenderer } = require("electron/renderer")
const img = await ipcRenderer.invoke("displays::capture-page");
```

This will return a valid NativeImage as expected. Tested on both Mac and Windows.
At least this solved my problem. I think this could be adapted to an alternative capturePage option.
